stdlib/public/core/Rotate.swift

```diff
@@ -0,0 +1,716 @@
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+extension MutableCollection {
+    /// Rotates the elements of the collection such that the value at the given
+    /// index moves leftward to `startIndex`.
+    ///
+    /// Passing `startIndex` or `endIndex` as `middle` has no effect.
+    ///
+    /// The following example rotates the elements of an array of characters so
+    /// the second character moves to be the first:
+    ///
+    ///     var café: [Character] = ["C", "a", "f", "é"]
+    ///     café.rotate(toFirst: café.index(after: café.startIndex))
+    ///     print(café)
+    ///     // Prints "["a", "f", "é", "C"]"
+    ///
+    /// - Precondition: `middle` must be a valid index of this collection.
+    ///
+    /// - Parameter middle: The index of the element whose value will move
+    ///   to `startIndex`, unless it's `endIndex` (then it's ignored).
+    ///
+    /// - Returns: The new index of the value originally at `startIndex`.
+    ///
+    /// - Postcondition: The collection is a left-rotation of its old state,
+    ///   *i.e.* `newValue` equals `oldValue[middle...] + oldValue[..<middle]`.
+    ///
+    /// - Complexity: O(*n*), where *n* is the number of elements in the
+    ///   collection.
+    @discardableResult
+    public mutating func rotate(toFirst middle: Index) -> Index {
+        switch middle {
+        case ...startIndex, endIndex...:
+            // A base case, otherwise the true/false and false/true cases of the
+            // switch below would trigger infinite recursion.
+            return startIndex
+        default:
+            // With the given index as the pivot between partitions, switch the
+            // order of said partitions.  Do it without testing the relative
+            // lengths in advance.  If the partitions have different lengths,
+            // the prefix/whole part of the originally-second partition is
+            // finalized and the remainder (the suffix of the longer original
+            // partition and the whole/prefix part of the originally-first
+            // partition) needs to be rotated.
+            var i = startIndex, j = middle
+            while true {
+                switch (i == middle, j == endIndex) {
+                case (false, false):
+                    // Move one step closer by doing the core swap.
+                    swapAt(i, j)
+                    formIndex(after: &i)
+                    formIndex(after: &j)
+                case (false, true):
+                    // The first partition is longer than the second.
+                    self[i...].rotate(toFirst: middle)
```

Comment:
Swift doesn't have guaranteed TCO, and IIUC the bound on call depth here is potentially O(n), so this should really be written in terms of a loop rather than recursively.

Personally, I would write this something like the following sketch; I find it to be clearer than the state machine on (Bool,Bool), but different people may have different opinions on matters of taste.
```
while j != endIndex {
  while i < middle && j < endIndex {
    /* swap, update indices */
  }
  // update i, j, middle, result for next pass
}
return result
```